proxy/src/connection.rs

```diff
@@ -185,23 +191,64 @@ impl Future for Connecting {
         loop {
             *self = match self {
-                Connecting::Plaintext { connect, tls } => {
+                Connecting::Plaintext { addr, connect, tls } => {
                     let plaintext_stream = try_ready!(connect.poll());
                     set_nodelay_or_warn(&plaintext_stream);
                     match tls.take().expect("Polled after ready") {
                         Conditional::Some(config) => {
-                            let upgrade_to_tls = tls::Connection::connect(
+                            let upgrading = tls::Connection::connect(
                                 plaintext_stream, &config.identity, config.config);
-                            Connecting::UpgradeToTls(upgrade_to_tls)
+                            Connecting::UpgradeToTls { addr: *addr, upgrading }
                         },
                         Conditional::None(why) => {
                             return Ok(Async::Ready(Connection::plain(plaintext_stream, why)));
                         },
                     }
                 },
-                Connecting::UpgradeToTls(upgrading) => {
-                    let tls_stream = try_ready!(upgrading.poll());
-                    return Ok(Async::Ready(Connection::tls(tls_stream)));
+                Connecting::UpgradeToTls { addr, upgrading } => {
+                    match upgrading.poll() {
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Ok(Async::Ready(tls_stream)) => {
+                            let conn = Connection::tls(tls_stream);
+                            return Ok(Async::Ready(conn));
+                        },
+                        Err(e) => {
+                            // Did the TLS handshake fail, or did something
+                            // else go wrong?
+                            let handshake_err = if let Some(tls_error) = e
+                                .get_ref()
+                                .and_then(|e| e.downcast_ref::<tls::Error>())
+                            {
+                                warn!(
+                                    "TLS handshake with {:?} failed: {}\
+                                     -> falling back to plaintext",
+                                    addr, tls_error,
+                                );
+                                true
+                            } else {
+                                false
+                            };
+                            if handshake_err {
+                                // XXX: We can't get the old connection back
+                                // from the failed upgrade future, so we have
+                                // to try reconnecting.
+                                let connect = TcpStream::connect(addr);
+                                // XXX: Should we propagate *why* the handshake
+                                // failed here?
```

Comment:
Ideally we would be more specific, but we can fix that in a follow-up.

More concerning to me is: how are we going to report this reason in telemetry? The reason we have a `ClientConfig` in this module instead of a `ClientConfigWatch` is that we had to extract the reason during binding, much earlier than now. Do we have a way to report telemetry based on information we gather this late?

Reply:
Any events on the new connection will have the `tls="handshake_failed"` label without any additional changes. Furthermore, I was thinking that (in a subsequent PR) we would change the `Connecting` future to generate a "handshake failed" event as well, which we could use to track a counter of handshake failures (that could be labeled with each individual error type).

Reply:
> Any events on the new connection will have the tls="handshake_failed" label without any additional changes. Furthermore, I was thinking that (in a subsequent PR) we would change the Connecting future to generate a "handshake failed" event as well, which we could use to track a counter of handshake failures (that could be labeled with each individual error type).

We have `TransportLabels::tls_status` too, which is set based on whether we'll *attempt* TLS, vs whether the resulting connection used TLS like here. In particular, `TransportLabels::tls_status` for outbound connections is set during binding, based on the value of `tls_status` from the `transport::Ctx`. Do we still need `TransportLabels::tls_status` given this? Or am I misunderstanding something?

Reply:
@briansmith that's not entirely correct --- we're still using `TransportLabels::tls_status` here, but in this case, when we fall back, the `tls_status` field is set to `HandshakeFailed`. 

The handshake failed _event_ I'm considering adding would be used for increment a separate counter of transport failures. I suppose we could also implement this counter on every `TransportOpen` event with the `HandshakeFailed` TLS status; maybe that's the better thing to do...